File: Java/src/ch08/TraverseBinaryTree.java
package ch08;

import java.util.ArrayList;

public class TraverseBinaryTree {

    public static class Node{
        public int value;
        public Node left;
        public Node right;

        public Node(int v){
            value = v;
        }
    }

    /**
     * 先序优先遍历
     * @param root 根节点
     * @return
     */
    public static ArrayList<Integer> pre(Node root){
        ArrayList<Integer> rnt = new ArrayList<>();
        if (root == null){
            return rnt;
        }
        rnt.add(root.value);
        rnt.addAll(pre(root.left));
        rnt.addAll(pre(root.right));
        return rnt;
    }

    /**
     * 中序优先遍历
     * @param root 根节点
     * @return
     */
    public static ArrayList<Integer> mid(Node root){
        ArrayList<Integer> rnt = new ArrayList<>();
        if (root == null){
            return rnt;
        }
        rnt.addAll(mid(root.left));
        rnt.add(root.value);
        rnt.addAll(mid(root.right));
        return rnt;
    }

    /**
     * 后序优先遍历
     * @param root 根节点
     * @return
     */
    public static ArrayList<Integer> pos(Node root){
        ArrayList<Integer> rnt = new ArrayList<>();
        if (root == null){
            return rnt;
        }
        rnt.addAll(pos(root.left));
        rnt.addAll(pos(root.right));
        rnt.add(root.value);
        return rnt;
    }


    public static void main(String[] args) {
        Node root = new Node(6);
        root.left = new Node(4);
        root.right = new Node(8);
        root.left.left = new Node(2);
        root.left.right = new Node(5);
        root.right.left = new Node(7);
        root.right.right = new Node(9);
        root.left.left.left = new Node(1);
        root.left.left.right = new Node(3);

        System.out.println("pre:");
        System.out.println(pre(root));
        System.out.println("mid:");
        System.out.println(mid(root));
        System.out.println("pos:");
        System.out.println(pos(root));
    }

}

File: Python/ch08/TraverseBinaryTree.py
# 二叉树的节点
class Node:
    def __init__(self, v):
        self.left = None
        self.right = None
        self.value = v


def pre(r):
    rnt = []
    if r is None:
        return rnt
    rnt.append(r.value)
    rnt.extend(pre(r.left))
    rnt.extend(pre(r.right))
    return rnt


def mid(r):
    rnt = []
    if r is None:
        return rnt
    rnt.extend(mid(r.left))
    rnt.append(r.value)
    rnt.extend(mid(r.right))
    return rnt


def pos(r):
    rnt = []
    if r is None:
        return rnt
    rnt.extend(pos(r.left))
    rnt.extend(pos(r.right))
    rnt.append(r.value)
    return rnt


if __name__ == '__main__':
    root = Node(6)
    root.left = Node(4)
    root.right = Node(8)
    root.left.left = Node(2)
    root.left.right = Node(5)
    root.right.left = Node(7)
    root.right.right = Node(9)
    root.left.left.left = Node(1)
    root.left.left.right = Node(3)

    print('pre:')
    print(pre(root))
    print('mid:')
    print(mid(root))
    print('pos:')
    print(pos(root))
